siliconcompiler/tools/yosys/syn_fpga.tcl

```diff
@@ -4,19 +4,157 @@ set job_name [dict get $sc_cfg option jobname]
+    #automated flows
+    #***NOTE:  because rename is a Tcl keyword the yosys -import
+    #          trick doesn't work
+    yosys rename -top ${sc_design}
+
+    yosys proc
+    #select -module ${sc_design}
+    #flatten -wb
+    #flatten
+
+    #Match the optimization passes of the VTR reference yosys flow
+    yosys opt_expr
+    yosys opt_clean
+    yosys check
+    yosys opt -nodffe -nosdff
+    yosys fsm
+    yosys opt
+    yosys wreduce
+    yosys peepopt
+    yosys opt_clean
+    yosys share
+    yosys opt
+    yosys memory -nomap
+    yosys opt -full
+
+    #Do our own thing from here
+    yosys techmap -map +/techmap.v
+
+    #***TO DO:  Figure out how to deal with this in an SC-compatible way
+    if { $sc_flop_techlib eq "None" } {
+	yosys dfflegalize -cell \$_DFF_P_ 01 -cell \$_DFF_P??_ 01 
+    }
+
+    #opt
+    #opt_clean
+
+    #Perform preliminary buffer insertion before passing to ABC to help reduce
+    #the overhead of final buffer insertion downstream
+    yosys insbuf
+
+    yosys abc -lut $sc_syn_lut_size -dff
+    #abc -script $env(FPGA_ARCHITECT_ROOT)/test/scripts/fpgaa_abc_flow.scr
+
+    #***TO DO:  Wrap this in a verbosity flag
+    #yosys write_verilog -noexpr -selected ${sc_design}_abc.v
+
+    #Clean up the design before buffer insertion to avoid blowing up the design
+    #size with that step
+    #opt
+    #clean
+
+    yosys flatten
+    yosys clean
+
+    #***NOTE:  If you flatten after this step instead of before Yosys barfs,
+    #          would be nice to have a deeper understanding of why that is
+
+    set sc_memmap_techlib [ dict get $sc_cfg tool $sc_tool task $sc_step var memmap ]
+    if { $sc_memmap_techlib ne "None" } {
+	yosys memory_libmap -lib $sc_memmap_techlib
+	#***HACK:  Eliminate $reduce_or with this mapping here
+	#          until we have a cleaner solution
+	yosys simplemap
+    }
+
+    #if { $map_buffers == 1 } {
+	#insbuf -buf $buffer_name $buf_input_name $buf_output_name
+    #}
+
+    #***NOTE:  hilomap needs to happen before tech mapping so that
+    #          we can run the hierarchy check and prevent the tiecell
+    #          library from being treated like part of the design
+
+    #if { $tiecells == 1 } {
+	#hilomap -hicell $tiehi_cell $tiehi_output_name -locell $tielo_cell $tielo_output_name
+    #}
+
+    ########################################################
+    # Technology Mapping
+    ########################################################
+    if [dict exists $sc_cfg tool $sc_tool var $sc_step $sc_index techmap ] {
+	set sc_techmap [ dict get $sc_cfg tool $sc_tool var $sc_step $sc_index techmap ]
+	# techmap_defines = [ dict get $sc_cfg tool $sc_tool var $sc_step $sc_index techmap_define ]
+	# techmap_define_args = ''
+	# foreach techmap_define $techmap_defines {
+	#    set techmap_define_args "$techmap_define_args -D$techmap_define"
+	# }
+	foreach mapfile $sc_techmap {
+	    # Add $techmap_define_args when ready
+	    yosys techmap -map $mapfile 
```

Comment:
In the asic flow we need to follow a techmap with:
https://github.com/siliconcompiler/siliconcompiler/blob/cebeb2f78bf4fc6498adba06aac80dcb702190b9/siliconcompiler/tools/yosys/syn_asic.tcl#L183C1-L188C2
In case this adds some unmapped logic like `assign x = ~b`

Reply:
Based on empirical evidence, the common FPGA case seems to be that abc mapping to LUTs addresses such issues.  I'm a bit concerned that there are corner cases I've overlooked where the ASIC flow analogy arises, but would need to run the flow on a large number of test cases to try and find a failure.